### ISSUE TITLE
Producer doesn't reconnect if broker goes down, reappears with new IP

### ISSUE THREAD

**Issue by originsmike** [dpkp/kafka-python#1083]
I'm using Docker swarm to manage Kafka brokers, so if a broker goes down, swarm will automatically create a new instance in new container. The catch is that the new container will likely have a different IP address from the old one. The issue is this: once that new broker comes up, kafka-python will never connect to it.

Here's an example running on my local machine, where I'm running a single node that can be discovered with 'kafka.local'. I run one Kafka broker which magically sets its advertised.listeners to the container's IP, which in this case was 172.17.0.4 for the old container and 172.17.0.5 for the new one. (both Kakfa and Zookeeper are in the same container, but I'm using a volume to persist their data so clients think it's the same broker when it finally comes back up)

Logs follow.

First, the node goes down and you see a lot of this:
```
2017-04-24T02:55:27.652Z conn.py:211 DEBUG:kafka.conn: <BrokerConnection host=172.17.0.4/172.17.0.4 port=9092>: creating new socket
2017-04-24T02:55:27.652Z sender.py:109 DEBUG:kafka.producer.sender: Node 0 not ready; delaying produce of accumulated batch
2017-04-24T02:55:27.653Z conn.py:295 ERROR:kafka.conn: Connect attempt to <BrokerConnection host=172.17.0.4/172.17.0.4 port=9092> returned error 111. Disconnecting.
2017-04-24T02:55:27.653Z client_async.py:304 WARNING:kafka.client: Node 0 connection failed -- refreshing metadata
2017-04-24T02:55:27.654Z conn.py:211 DEBUG:kafka.conn: <BrokerConnection host=172.17.0.4/172.17.0.4 port=9092>: creating new socket
2017-04-24T02:55:27.655Z sender.py:109 DEBUG:kafka.producer.sender: Node 0 not ready; delaying produce of accumulated batch
2017-04-24T02:55:27.655Z conn.py:295 ERROR:kafka.conn: Connect attempt to <BrokerConnection host=172.17.0.4/172.17.0.4 port=9092> returned error 111. Disconnecting.
2017-04-24T02:55:27.656Z client_async.py:304 WARNING:kafka.client: Node 0 connection failed -- refreshing metadata
2017-04-24T02:55:27.656Z client_async.py:656 ERROR:kafka.client: No nodes found in metadata -- retrying bootstrap
2017-04-24T02:55:27.656Z client_async.py:213 DEBUG:kafka.client: Sleeping 12.6888 before bootstrapping again
```
Meanwhile, we send more messages:
```
2017-04-24T02:55:30.202Z kafka.py:508 DEBUG:kafka.producer.kafka: Sending (key=1 value=1) to TopicPartition(topic='_test-count', partition=0)
```
kafka-python realizes something's terribly wrong and tries bootstrapping again using the original bootstrap_server value, which is good:
```
2017-04-24T02:55:40.355Z client_async.py:223 DEBUG:kafka.client: Attempting to bootstrap via node at kafka.local:9092
2017-04-24T02:55:40.357Z conn.py:211 DEBUG:kafka.conn: <BrokerConnection host=kafka.local/kafka.local port=9092>: creating new socket
2017-04-24T02:55:40.358Z conn.py:281 DEBUG:kafka.conn: <BrokerConnection host=kafka.local/127.0.0.1 port=9092>: established TCP connection
2017-04-24T02:55:40.359Z client_async.py:274 DEBUG:kafka.client: Node bootstrap connected
2017-04-24T02:55:40.362Z conn.py:545 DEBUG:kafka.conn: <BrokerConnection host=kafka.local/127.0.0.1 port=9092> Request 1: MetadataRequest_v1(topics=NULL)
2017-04-24T02:55:40.387Z conn.py:721 DEBUG:kafka.conn: <BrokerConnection host=kafka.local/127.0.0.1 port=9092> Response 1: MetadataResponse_v1(brokers=[(node_id=0, host='172.17.0.5', port=9092, rack=None)], controller_id=0, topics=[(error_code=0, topic='_test-count', is_internal=False, partitions=[(error_code=0, partition=0, leader=0, replicas=[0], isr=[0])]), (error_code=0, topic='aqi-ground-station-raw-data', is_internal=False, partitions=[(error_code=0, partition=0, leader=0, replicas=[0], isr=[0])])])
2017-04-24T02:55:40.388Z cluster.py:272 DEBUG:kafka.cluster: Updated cluster metadata to ClusterMetadata(brokers: 1, topics: 2, groups: 0)
```
It's successful, but the client object doesn't seem to notice:
```
2017-04-24T02:55:40.389Z client_async.py:707 DEBUG:kafka.client: Give up sending metadata request since no node is available
```
And continues doing this forever:
```
2017-04-24T02:55:40.391Z conn.py:211 DEBUG:kafka.conn: <BrokerConnection host=172.17.0.4/172.17.0.4 port=9092>: creating new socket
2017-04-24T02:55:40.392Z sender.py:109 DEBUG:kafka.producer.sender: Node 0 not ready; delaying produce of accumulated batch
2017-04-24T02:55:40.392Z conn.py:295 ERROR:kafka.conn: Connect attempt to <BrokerConnection host=172.17.0.4/172.17.0.4 port=9092> returned error 111. Disconnecting.
2017-04-24T02:55:40.393Z client_async.py:304 WARNING:kafka.client: Node 0 connection failed -- refreshing metadata
2017-04-24T02:55:40.394Z conn.py:211 DEBUG:kafka.conn: <BrokerConnection host=172.17.0.4/172.17.0.4 port=9092>: creating new socket
2017-04-24T02:55:40.395Z sender.py:109 DEBUG:kafka.producer.sender: Node 0 not ready; delaying produce of accumulated batch
2017-04-24T02:55:40.396Z conn.py:295 ERROR:kafka.conn: Connect attempt to <BrokerConnection host=172.17.0.4/172.17.0.4 port=9092> returned error 111. Disconnecting.
```
Another variation of this bug happens when, during bootstrap, the name in bootstrap_servers fails to resolve (socket.gaierror) and comes back up later. The reconnect logic seems complicated though, so I doubt the same fix would work for that.

This is a ship stopper for me (and should be for anyone else who's running Kakfa in docker swarm) so I'm going to start digging in and come up with a fix. Any help would be appreciated though!

**Comment by dpkp** [comment]
Are you testing with kafka-python master?

**Comment by originsmike** [comment]
Looks like KafkaClient._maybe_connect is the right place to make the fix. In that function, once a connection for a given node_id is created, it'll never be updated if the host IP or post changes.

Made a quick fix that does a quick check to see if the metadata from the broker changed, and if so, shuts down the existing connection and allows a new one to be created. Seems to work.

I'll do some testing with the more extreme version of this scenario (bootstrap_servers name fails to resolve for a few seconds) and hopefully it'll fix that as well.

**Comment by originsmike** [comment]
1.3.2. Sorry I didn't mention this before. Is this fixed in master?

**Comment by jbreis** [comment]
Hi guys,
Do you know if this issue is also fixed on java producer client ?

Thanks

**Comment by jeffwidman** [comment]
@jbreis afraid I don't know, that would be better asked on StackOverflow or the Kafka mailing list and keep this ticket focused on the issue in `kafka-python`

**Comment by dpkp** [comment]
This issue was fixed in PR #1085
